godeltaprof/internal/pprof/delta_heap.go

```diff
@@ -24,26 +49,30 @@ func (d *DeltaHeapProfiler) WriteHeapProto(b ProfileBuilder, p []runtime.MemProf
 		if r.AllocObjects > 0 {
 			blockSize = r.AllocBytes / r.AllocObjects
 		}
-		entry := d.m.Lookup(r.Stack(), uintptr(blockSize))
+		entry := d.m.Lookup(stack(r.Stack0[:]), uintptr(blockSize))
+		if entry.acc == (heapAccValue{}) {
+			continue
+		}
 
-		if (r.AllocObjects - entry.count.v1) < 0 {
+		AllocObjects := entry.acc.allocObjects - entry.prev.allocObjects
+		if AllocObjects < 0 {
 			continue
 		}
-		AllocObjects := r.AllocObjects - entry.count.v1
-		AllocBytes := r.AllocBytes - entry.count.v2
-		entry.count.v1 = r.AllocObjects
-		entry.count.v2 = r.AllocBytes
+		AllocBytes := AllocObjects * blockSize
+		entry.prev.allocObjects = entry.acc.allocObjects
 
-		values[0], values[1] = scaleHeapSample(AllocObjects, AllocBytes, rate)
-		values[2], values[3] = scaleHeapSample(r.InUseObjects(), r.InUseBytes(), rate)
+		values[0], values[1] = ScaleHeapSample(AllocObjects, AllocBytes, rate)
+		values[2], values[3] = ScaleHeapSample(entry.acc.inuseObjects, entry.acc.inuseObjects*blockSize, rate)
+
+		entry.acc = heapAccValue{}
```

Comment:
This change is not very clear to me – could you please add a comment why we're doing this?

Also, I'm a bit concerned about things like:
```
AllocBytes := AllocObjects * blockSize
```
and
```
entry.acc.inuseObjects*blockSize // inuse bytes
```

Shouldn't we use the values from the `MemProfileRecord`?

Reply:
I've added this comment to the source code. 

```
		// allocBytes, inuseBytes is calculated as multiplication of number of objects by blockSize
		// This is done to reduce the size of the map entry (i.e. heapAccValue for deduplication and
		// heapPrevValue for keeping the delta).
```
Bascialy I decided to rely on yet another implementation detail,  the fact that `allocSpace` and `inuseSpace` is always multiple of `blockSize` in the current golang runtime implementation. But this may change in the future.
So i see two options here: rely on this implementation detail or keep 3 extra fields for each heap record
Does the comment make sense?
Should we keep relying on this or better make it more reliable for the future?

Reply:
Thank you for the explanation, this does make sense! I don't have a strong opinion on this